community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/state/TransactionRecordState.java

```diff
@@ -135,59 +131,72 @@ public void extractCommands( Collection<StorageCommand> commands ) throws Transa
 
         integrityValidator.validateTransactionStartKnowledge( lastCommittedTxWhenTransactionStarted );
 
-        int noOfCommands = context.getNodeRecords().changeSize() +
-                           context.getRelRecords().changeSize() +
-                           context.getPropertyRecords().changeSize() +
-                           context.getSchemaRuleChanges().changeSize() +
-                           context.getPropertyKeyTokenRecords().changeSize() +
-                           context.getLabelTokenRecords().changeSize() +
-                           context.getRelationshipTypeTokenRecords().changeSize() +
-                           context.getRelGroupRecords().changeSize() +
+        int noOfCommands = recordChangeSet.changeSize() +
                            (neoStoreRecord != null ? neoStoreRecord.changeSize() : 0);
 
-        for ( RecordProxy<Integer, LabelTokenRecord, Void> record : context.getLabelTokenRecords().changes() )
+        for ( RecordProxy<Integer, LabelTokenRecord, Void> record : recordChangeSet.getLabelTokenChanges().changes() )
         {
             commands.add( new Command.LabelTokenCommand( record.getBefore(), record.forReadingLinkage() ) );
         }
-        for ( RecordProxy<Integer, RelationshipTypeTokenRecord, Void> record : context.getRelationshipTypeTokenRecords().changes() )
+        for ( RecordProxy<Integer, RelationshipTypeTokenRecord, Void> record :
+            recordChangeSet.getRelationshipTypeTokenChanges().changes() )
         {
             commands.add( new Command.RelationshipTypeTokenCommand( record.getBefore(), record.forReadingLinkage() ) );
         }
-        for ( RecordProxy<Integer, PropertyKeyTokenRecord, Void> record : context.getPropertyKeyTokenRecords().changes() )
+        for ( RecordProxy<Integer, PropertyKeyTokenRecord, Void> record :
+            recordChangeSet.getPropertyKeyTokenChanges().changes() )
         {
             commands.add( new Command.PropertyKeyTokenCommand( record.getBefore(), record.forReadingLinkage() ) );
         }
 
         // Collect nodes, relationships, properties
-        List<Command> nodeCommands = new ArrayList<>( context.getNodeRecords().changeSize() );
-        for ( RecordProxy<Long, NodeRecord, Void> change : context.getNodeRecords().changes() )
+        List<Command> nodeCommands = Collections.emptyList();
```

Comment:
We can even start using arrays instead of these `ArrayList`s since we always know the size.